### PR TITLE
perf: Add metric for time spent casting in native scan

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

Make it easy to see how much of the native `ScanExec` time is spent casting columns to different types (this usually means unpacking dictionaries). 

Example from TPC-DS q9:

Screenshot from 2024-09-06 11-57-48

DataFusion metrics in native explain output:

```
metrics=[
  output_rows=2097152, 
  elapsed_compute=21.847892ms, 
  cast_time=21.631731ms]
```

Full plan:

```
AggregateExec: mode=Partial, gby=[], aggr=[count, avg, avg], metrics=[output_rows=1, elapsed_compute=9.481194ms]
  ProjectionExec: expr=[col_1@1 as col_0, col_2@2 as col_1], metrics=[output_rows=400519, elapsed_compute=50.596µs]
    FilterExec: col_0@0 IS NOT NULL AND col_0@0 >= 81 AND col_0@0 <= 100, metrics=[output_rows=400519, elapsed_compute=4.753725ms]
      ScanExec: source=[CometScan parquet  (unknown)], schema=[col_0: Int32, col_1: Decimal128(7, 2), col_2: Decimal128(7, 2)], metrics=[output_rows=2097152, elapsed_compute=21.847892ms, cast_time=21.631731ms]
```

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/comet/CometMetricNode.scala [hunk 1]
```diff
@@ -79,6 +79,15 @@ object CometMetricNode {
         "total time (in ms) spent in this operator"))
   }
 
+  /**
+   * SQL Metrics for Comet native ScanExec
+   */
+  def scanMetrics(sc: SparkContext): Map[String, SQLMetric] = {
+    Map(
+      "cast_time" ->
+        SQLMetrics.createNanoTimingMetric(sc, "Total time for casting arrays"))
```
Comment (viirya) [inline spark/src/main/scala/org/apache/spark/sql/comet/CometMetricNode.scala:88]: "arrays" in "casting arrays" may be confused with Spark array type and considered as time spent on casting (Spark) "arrays".

Maybe "casting columns"?

File: native/core/src/execution/operators/scan.rs [hunk 2]
```diff
@@ -345,16 +347,20 @@ struct ScanStream<'a> {
     baseline_metrics: BaselineMetrics,
     /// Cast options
     cast_options: CastOptions<'a>,
+    /// Timer for cast operations
```
Comment (comphead) [inline native/core/src/execution/operators/scan.rs:350]: is this elapsed time?

File: spark/src/main/scala/org/apache/spark/sql/comet/CometScanExec.scala [hunk 3]
```diff
@@ -198,9 +198,12 @@ case class CometScanExec(
     // Tracking scan time has overhead, we can't afford to do it for each row, and can only do
     // it for each batch.
     if (supportsColumnar) {
-      Some("scanTime" -> SQLMetrics.createNanoTimingMetric(sparkContext, "scan time"))
+      Seq(
```
Comment (comphead) [inline spark/src/main/scala/org/apache/spark/sql/comet/CometScanExec.scala:201]: should be it a Map instead of Seq?

**Comment by andygrove** [comment]
@viirya @comphead I have addressed feedback. Thanks.

**Review by viirya — APPROVED** [review]

**Review by comphead — APPROVED** [review]
lgtm thanks @andygrove

**Pull request merged**